File: telegram-service/src/main/java/de/nak/home_assistant/Bot.java
package de.nak.home_assistant;

import com.pengrad.telegrambot.TelegramBot;
import com.pengrad.telegrambot.model.BotCommand;
import com.pengrad.telegrambot.model.Update;
import com.pengrad.telegrambot.request.SetMyCommands;
import de.nak.home_assistant.commands.*;
import de.nak.home_assistant.models.telegram.Command;

import java.util.Arrays;
import java.util.List;

public class Bot {

    TelegramBot api = new TelegramBot(Env.getBotToken());
    public static List<Command> FIXED_COMMANDS = Arrays.asList(new Start(), new Settings());

    public Bot() {
        BotCommand[] commands = FIXED_COMMANDS.stream().map(c -> new BotCommand(c.getName(), c.getDescription()))
                .toArray(BotCommand[]::new);

        SetMyCommands setMyCommands = new SetMyCommands(commands);
        api.execute(setMyCommands);
    }

    public void handleUpdate(Update update) {

        if (update.callbackQuery() != null) {
            Callback callback = new Callback(api);
            callback.replyToButtons(update.callbackQuery());
            return;
        }

        if (update.message().replyToMessage() != null) {
            Input inputCommand = new Input(api);
            inputCommand.replyToReply(update.message());
            return;
        }

        handleCommand(update);
    }

    private boolean commandMatcher(String inputText, Command handler) {
        String input = inputText.toLowerCase();
        return input.startsWith(handler.getName()) || input.startsWith("/" + handler.getName());
    }

    private void handleCommand(Update update) {
        Fallback fallback = new Fallback(api);

        if (update.message() == null) {
            fallback.reply(update);
            return;
        }

        String text = update.message().text();
        Command[] commands = { new Start(api), new Settings(api), new Lifx(api), new Hue(api), new Fritz(api) };

        Command command = Arrays.stream(commands).filter(handler -> commandMatcher(text, handler)).findFirst()
                .orElse(fallback);

        command.reply(update);
    }
}

File: telegram-service/src/bot.ts
import TelegramBot from "node-telegram-bot-api";
import { replyToButtons } from "./commands/callback";
import { Fritz, replyToFritz } from "./commands/fritz";
import { Hue, replyToHue } from "./commands/hue";
import { replyToReply } from "./commands/input";
import { Lifx, replyToLifx } from "./commands/lifx";
import {
  ModuleSettings,
  replyToModuleSettins,
  replyToSettings,
  replyToUserSettings,
  Settings,
  UserSettings,
} from "./commands/settings";
import { Back, replyToBack, replyToStart, Start } from "./commands/start";
import { callback } from "./handler";

const { BOT_TOKEN } = process.env;

export const FIXED_COMMANDS = [Start, Settings];

export const bot = new TelegramBot(BOT_TOKEN, {
  webHook: false,
  polling: false,
});

export const generateBot = () => {
  bot.clearTextListeners();

  bot.removeAllListeners("callback_query");
  bot.removeAllListeners("pinned_message");

  bot.onText(new RegExp(`/?${Start().command}`), async (msg) => {
    await replyToStart(msg).catch((e) =>
      bot.sendMessage(msg.chat.id, `start: ${e}`)
    );
    callback();
  });

  bot.onText(new RegExp(`/?${Back().command}`), async (msg) => {
    await replyToBack(msg).catch((e) => bot.sendMessage(msg.chat.id, e));
    callback();
  });

  bot.onText(new RegExp(`/?${Settings().command}`), async (msg) => {
    await replyToSettings(msg).catch((e) =>
      bot.sendMessage(msg.chat.id, `einstellungen: ${e}`)
    );
    callback();
  });

  bot.onText(new RegExp(`/?${ModuleSettings().command}`), async (msg) => {
    await replyToModuleSettins(msg).catch((e) =>
      bot.sendMessage(msg.chat.id, `module: ${e}`)
    );
    callback();
  });

  bot.onText(new RegExp(`/?${UserSettings().command}`), async (msg) => {
    await replyToUserSettings(msg).catch((e) =>
      bot.sendMessage(msg.chat.id, `benutzer: ${e}`)
    );
    callback();
  });

  bot.onText(new RegExp(`/?${Lifx().command} ?(.+)?`), async (msg, match) => {
    await replyToLifx(msg, match).catch((e) =>
      bot.sendMessage(msg.chat.id, `lifx: ${e}`)
    );
    callback();
  });

  bot.onText(new RegExp(`/?${Hue().command}`), async (msg) => {
    await replyToHue(msg).catch((e) =>
      bot.sendMessage(msg.chat.id, `hue: ${e}`)
    );
    callback();
  });

  bot.onText(new RegExp(`/?${Fritz().command} ?(.+)?`), async (msg, match) => {
    await replyToFritz(msg, match).catch((e) =>
      bot.sendMessage(msg.chat.id, `fritz: ${e}`)
    );
    callback();
  });

  bot.onText(new RegExp(""), async (msg) => {
    if (!msg.reply_to_message) return;

    await replyToReply(msg).catch((e) =>
      bot.sendMessage(msg.chat.id, `text: ${e}`)
    );
    callback();
  });

  bot.on("callback_query", async (cbQuery) => {
    await replyToButtons(cbQuery).catch((e) =>
      bot.sendMessage(cbQuery.message.chat.id, `callback_query: ${e}`)
    );
    callback();
  });

  bot.on("pinned_message", () => callback());

  return bot;
};
